be/src/olap/memory/typed_column_writer.h

```diff
@@ -0,0 +1,305 @@
+        RETURN_IF_ERROR(block->alloc(Column::BLOCK_SIZE, sizeof(ST)));
+        _base->emplace_back(std::move(block));
+        if (_column->schema().cid() == 1) {
+            // only log when first column add block
+            DLOG(INFO) << StringPrintf("Column(cid=%u) add ColumnBlock %zu/%zu",
+                                       _column->schema().cid(), _base->size(), _base->capacity());
+        }
+        return Status::OK();
+    }
+
+    // Expand versions vector, do a copy-on-write
+    Status expand_versions() {
+        size_t new_capacity =
+                padding(_column->_versions.capacity() + Column::VERSION_CAPACITY_STEP_SIZE,
+                        Column::VERSION_CAPACITY_STEP_SIZE);
+        DLOG(INFO) << StringPrintf("%s memory=%.1lfM expand delta base=%zu version=%zu",
+                                   _column->schema().debug_string().c_str(),
+                                   _column->memory() / 1000000.0, _base->capacity(), new_capacity);
+        scoped_refptr<Column> cow(new Column(*_column.get(), 0, new_capacity));
+        cow.swap(_column);
+        return Status::OK();
+    }
+
+    // Try to add a new delta to versions vector, if versions vector is at it's capcacity,
+    // call expand versions
+    Status add_delta(scoped_refptr<ColumnDelta>&& delta, uint64_t version) {
+        if (_column->_versions.size() == _column->_versions.capacity()) {
+            expand_versions();
+        }
+        DLOG(INFO) << StringPrintf("%s add version %zu update: %zu",
+                                   _column->debug_string().c_str(), version, delta->size());
+        CHECK_LT(_column->_versions.size(), _column->_versions.capacity());
+        _column->_versions.emplace_back();
+        Column::VersionInfo& vinfo = _column->_versions.back();
+        vinfo.version = version;
+        vinfo.delta = delta;
+        return Status::OK();
+    }
+
+    template <class Reader, class T2, bool Nullable2, class ST2>
+    friend const void* TypedColumnGet(const Reader& reader, const uint32_t rid);
+
+    template <class T2, class ST2>
+    friend bool TypedColumnHashcode(const void*, size_t);
+
+    template <class Reader, class T2, bool Nullable2, class ST2>
+    friend bool TypedColumnEquals(const Reader&, const uint32_t, const void*, size_t);
+
+    // The following members need to be identical to TypedColumnReader
+    // because they share same reader methods: get/hashcode/equals
+    scoped_refptr<Column> _column;
+    vector<scoped_refptr<ColumnBlock>>* _base;
+    uint64_t _version;
+    uint64_t _real_version;
+    vector<ColumnDelta*> _deltas;
+
+    size_t _num_insert = 0;
+    size_t _num_update = 0;
+    bool _update_has_null = false;
+    typedef typename std::conditional<Nullable, NullableUpdateType<UT>, UpdateType<UT>>::type
+            UpdateMapType;
+    // Temporary stoarage to hold all updated cells' values
+    // rowid -> updated values
+    std::map<uint32_t, UpdateMapType> _updates;
```

Comment:
why not use unordered_map?

Reply:
updates will be finalized to delta, and delta_index needs to be in order, std::map has order.